be2-scala/src/main/scala/ch/epfl/pop/storage/DiskStorage.scala

```diff
@@ -0,0 +1,62 @@
+package ch.epfl.pop.storage
+
+import java.io.{File, IOException}
+import java.nio.charset.StandardCharsets
+
+import org.iq80.leveldb.{DB, Options, WriteBatch}
+import org.iq80.leveldb.impl.Iq80DBFactory.factory
+import org.slf4j.{Logger, LoggerFactory}
+
+class DiskStorage(val databaseFolder: String = DiskStorage.DATABASE_FOLDER) {
+
+  // create a logger for the database system
+  val logger: Logger = LoggerFactory.getLogger("DbLogger")
+
+  val options: Options = new Options()
+  options.createIfMissing(true)
+  options.cacheSize(DiskStorage.CACHE_SIZE)
+
+  val db: DB = {
+    try factory.open(new File(databaseFolder), options)
+    catch {
+      case e: IOException =>
+        logger.error("Could not open database folder '{}': {}", databaseFolder, e.getMessage)
+        throw e
+    }
+  }
+
+  def close(): Unit = db.close()
+
+
+  @throws [org.iq80.leveldb.DBException]
+  def read(key: String): String = {
```

Comment:
would it make sense to log & return an Option instead of throwing ? The higher layers probably won't be able to deal with the error anyways.

Reply:
No the whole new actor 10x shorter than the previous one (in part) because of this ^^. Let me show you how it works and we can then discuss if it's a good or if a change would be better :)

<img width="878" alt="Screenshot 2022-02-14 at 15 18 23" src="https://user-images.githubusercontent.com/44499915/153881055-299f7926-fa78-4895-b56b-fd80ac1b8de2.png">

Look at the `write` method. We spare 2 levels of indentation (match on Option & Try) using the fact that IF the readChannelData fails, then the rest of the method (the storage.write call) will not be execute and an exception will be thrown.

This alone is not enough to justify the use of Exceptions I agree. However here's the catch! For Tuomas to use his futures correctly, the dbActor needs to send back Futures (either successful containing an ACK, or failure containing an exception). HOWEVER, it turns out that an actor does *not* send back a failed future on exception throw (didn't know that ^^, thank you Tuomas for the enlightenment). This is why we have to have a Try block in the actor before sending back to client.

It is not there yet because it is being implemented in Tuomas PR, but you can see the screenshot above for an example (line 23 and 24)

Reply:
:+1: